File: Wire-iOS/Sources/UserInterface/Helpers/UIView+Zeta.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "UIView+Zeta.h"
#import "UIResponder+FirstResponder.h"
#import "Wire-Swift.h"

static NSString * const WireLastCachedKeyboardHeightKey = @"WireLastCachedKeyboardHeightKey";

@implementation UIView (Zeta)

+ (void)animateWithKeyboardNotification:(NSNotification *)notification
                                 inView:(UIView *)view
                             animations:(void (^)(CGRect keyboardFrameInView))animations
                             completion:(void (^)(BOOL finished))completion
{
    [self animateWithKeyboardNotification:notification inView:view delay:0 animations:animations completion:completion];
}

+ (void)animateWithKeyboardNotification:(NSNotification *)notification
                                 inView:(UIView *)view
                                  delay:(NSTimeInterval)delay
                             animations:(void (^)(CGRect keyboardFrameInView))animations
                             completion:(void (^)(BOOL finished))completion
{
    CGRect keyboardFrame = [self keyboardFrameInView:view forKeyboardNotification:notification];
    
    UIResponder *currentFirstResponder = [UIResponder wr_currentFirstResponder];
    if (currentFirstResponder != nil) {
        CGSize keyboardSize = CGSizeMake(keyboardFrame.size.width, keyboardFrame.size.height - currentFirstResponder.inputAccessoryView.bounds.size.height);
        [self wr_setLastKeyboardSize:keyboardSize];
    }
    
    NSDictionary *userInfo = notification.userInfo;
    double animationLength = [[userInfo objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
    UIViewAnimationCurve const animationCurve = (UIViewAnimationCurve) [userInfo[UIKeyboardAnimationCurveUserInfoKey] unsignedIntegerValue];
    
    UIViewAnimationOptions animationOptions = UIViewAnimationOptionBeginFromCurrentState;
    if (animationCurve == UIViewAnimationCurveEaseIn) {
        animationOptions |= UIViewAnimationOptionCurveEaseIn;
    }
    else if (animationCurve == UIViewAnimationCurveEaseInOut) {
        animationOptions |= UIViewAnimationOptionCurveEaseInOut;
    }
    else if (animationCurve == UIViewAnimationCurveEaseOut) {
        animationOptions |= UIViewAnimationOptionCurveEaseOut;
    }
    else if (animationCurve == UIViewAnimationCurveLinear) {
        animationOptions |= UIViewAnimationOptionCurveLinear;
    }
    
    [UIView animateWithDuration:animationLength
                          delay:delay
                        options:animationOptions
                     animations:^{
                         animations(keyboardFrame);
                     }
                     completion:completion];
}

+ (void)wr_setLastKeyboardSize:(CGSize)lastSize
{
    [[NSUserDefaults standardUserDefaults] setObject:NSStringFromCGSize(lastSize) forKey:WireLastCachedKeyboardHeightKey];
}

+ (CGSize)wr_lastKeyboardSize
{
    NSString *currentLastValue = [[NSUserDefaults standardUserDefaults] objectForKey:WireLastCachedKeyboardHeightKey];
    
    if (currentLastValue == nil) {
        return CGSizeMake([UIScreen mainScreen].bounds.size.width, KeyboardHeight.current);
    }
    else {
        CGSize keyboardSize = CGSizeFromString(currentLastValue);
        
        // If keyboardSize value is clearly off we need to pull default value
        if (keyboardSize.height < 150) {
            keyboardSize.height = KeyboardHeight.current;
        }
        
        return keyboardSize;
    }
}

+ (CGRect)keyboardFrameInView:(UIView *)view forKeyboardNotification:(NSNotification *)notification
{
    NSDictionary *userInfo = notification.userInfo;
    return [self keyboardFrameInView:view forKeyboardInfo:userInfo];
}

+ (CGRect)keyboardFrameInView:(UIView *)view forKeyboardInfo:(NSDictionary *)keyboardInfo
{
    CGRect screenRect = [[keyboardInfo objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
    CGRect windowRect = [view.window convertRect:screenRect fromWindow:nil];
    CGRect viewRect = [view convertRect:windowRect fromView:nil];
    
    CGRect intersection = CGRectIntersection(viewRect, view.bounds);
    
    return intersection;
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/Helpers/UIView+Zeta.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit

private let WireLastCachedKeyboardHeightKey = "WireLastCachedKeyboardHeightKey"

extension UIView {
    /// Provides correct handling for animating alongside a keyboard animation
    class func animate(
        withKeyboardNotification notification: Notification?,
        in view: UIView,
        delay: TimeInterval = 0,
        animations: @escaping (_ keyboardFrameInView: CGRect) -> Void,
        completion: ResultHandler? = nil
    ) {
        let keyboardFrame = keyboardFrame(in: view, forKeyboardNotification: notification)

        if let currentFirstResponder = UIResponder.currentFirst {
            let keyboardSize = CGSize(
                width: keyboardFrame.size.width,
                height: keyboardFrame.size.height - (currentFirstResponder.inputAccessoryView?.bounds.size.height ?? 0)
            )
            setLastKeyboardSize(keyboardSize)
        }

        let userInfo = notification?.userInfo
        let animationLength: TimeInterval = (userInfo?[UIResponder.keyboardAnimationDurationUserInfoKey] as? NSNumber)?
            .doubleValue ?? 0
        let animationCurve =
            AnimationCurve(
                rawValue: (userInfo?[UIResponder.keyboardAnimationCurveUserInfoKey] as AnyObject)
                    .intValue ?? 0
            ) ?? .easeInOut

        var animationOptions: UIView.AnimationOptions = .beginFromCurrentState

        switch animationCurve {
        case .easeIn:
            animationOptions.insert(.curveEaseIn)
        case .easeInOut:
            animationOptions.insert(.curveEaseInOut)
        case  .easeOut:
            animationOptions.insert(.curveEaseOut)
        case  .linear:
            animationOptions.insert(.curveLinear)
        default:
            break
        }

        UIView.animate(withDuration: animationLength, delay: delay, options: animationOptions, animations: {
            animations(keyboardFrame)
        }, completion: completion)
    }

    class func setLastKeyboardSize(_ lastSize: CGSize) {
        UserDefaults.standard.set(NSCoder.string(for: lastSize), forKey: WireLastCachedKeyboardHeightKey)
    }

    class var lastKeyboardSize: CGSize {
        if let currentLastValue = UserDefaults.standard.object(forKey: WireLastCachedKeyboardHeightKey) as? String {
            var keyboardSize = NSCoder.cgSize(for: currentLastValue)

            // If keyboardSize value is clearly off we need to pull default value
            if keyboardSize.height < 150 {
                keyboardSize.height = KeyboardHeight.current
            }

            return keyboardSize
        }

        return CGSize(width: UIScreen.main.bounds.size.width, height: KeyboardHeight.current)
    }

    class func keyboardFrame(in view: UIView, forKeyboardNotification notification: Notification?) -> CGRect {
        let userInfo = notification?.userInfo
        return keyboardFrame(in: view, forKeyboardInfo: userInfo)
    }

    class func keyboardFrame(in view: UIView, forKeyboardInfo keyboardInfo: [AnyHashable: Any]?) -> CGRect {
        let screenRect = keyboardInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect
        let windowRect = view.window?.convert(screenRect ?? CGRect.zero, from: nil)
        let viewRect = view.convert(windowRect ?? CGRect.zero, from: nil)

        return viewRect.intersection(view.bounds)
    }
}
